File: services/learn/learn/domain/users/user_service.py
from learn.infrastructure.repositories.user_repository import user_repository


class UserService:
    def __init__(self, user_repository):
        self.user_repository = user_repository

    def get_users(self):
        users = self.user_repository.get_users()
        return users

    def get_user_by_id(self, id):
        user = self.user_repository.get_user_by_id(id)
        return user

    def create_user(self, name, email):
        user = self.user_repository.create_user(name, email)
        return user

    def update_user(self, id, name, email):
        user = self.user_repository.update_user(id, name, email)
        return user

    def delete_user_by_id(self, id):
        self.user_repository.delete_user_by_id(id)


user_service = UserService(user_repository)

File: services/Users/Services/Users/UserService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Users.BuildingBlocks.Security;
using Users.Entities;
using Users.Infrastructure.EntityFramework.Repositories.Users;
using Users.Services.Users.IntegrationEvents;

namespace Users.Services.Users
{
    public class UserService
    {
        private readonly IUserRepository _userRepository;
        private readonly ISecurityService _securityService;
        private readonly UsersIntegrationEventsPublisher _usersIntegrationEventsPublisher;

        public UserService(IUserRepository userRepository, ISecurityService securityService, UsersIntegrationEventsPublisher usersIntegrationEventsPublisher)
        {
            _userRepository = userRepository;
            _securityService = securityService;
            _usersIntegrationEventsPublisher = usersIntegrationEventsPublisher;
        }

        public async Task<IList<UserDto>> GetAllUsers()
        {
            return (await _userRepository.GetAll())
                .Select(u => new UserDto()
                {
                    Id = u.Id,
                    Name = u.Name,
                    Email = u.Email,
                    Roles = u.Roles
                        .Select(r => new RoleDto() { Id = r.Id, Name = r.Name })
                        .ToList()
                })
                .ToList();
        }

        public async Task CreateUser(
            Guid id,
            string email,
            string name,
            string password,
            List<Role> roles)
        {
            var user = User.CreateNew(
                email,
                name,
                password,
                roles,
                _securityService);

            user.Id = id;

            await _userRepository.CreateUser(user);

            _usersIntegrationEventsPublisher.PublishUserCreatedEvent(new UserCreatedIntegrationEvent(user.Id, user.Name, user.Email, roles.Select(r => r.Name).ToList()));
        }
    }
}
